clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java

```diff
@@ -122,8 +124,8 @@ public final ByteBuffer serializeWithHeader(RequestHeader header) {
     }
 
     // Visible for testing
-    public final ByteBuffer serialize() {
-        return MessageUtil.toByteBuffer(data(), version);
+    public final Readable serialize() {
```

Comment:
Maybe just return `ByteAccessor`?

Reply:
> Maybe just return `ByteAccessor`?

Added an overloaded method for test cases where they need ByteBuffer to validate properties.

Reply:
I was kind of thinking maybe it should just return `ByteAccessor` without having another overload?  All the code that takes `Readable` would just work, and a small number of places that need to work with the buffer could use the result as the `ByteAccessor`.  Pretty much similar to what you had with downcasting only without downcasting :-).

Reply:
Then probably it would make sense to make `toByteBufferAccessor` public, and I'm not sure if we need `toReadable` and `toByteBuffer` overloads as the data can be easily extracted from `ByteBufferAccessor`.